internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,385 @@
+// in golang.org/x/sys/windows is located in /internal/winapi.
+//
+// https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects
+
+// JobObject is a high level wrapper around a Windows job object. Holds a handle to
+// the job, a channel to receive iocp notifications about the lifecycle
+// of the job and a mutex for synchronized handle access.
+type JobObject struct {
+	jobHandle  windows.Handle
+	notifChan  chan interface{}
+	handleLock sync.RWMutex
+}
+
+type jobLimits struct {
+	cpuCount        uint32
+	cpuRate         uint32
+	cpuWeight       uint32
+	memoryLimitInMB uintptr
+	maxIops         int64
+	maxBandwidth    int64
+}
+
+// Processor resource controls
+const (
+	processorLimitMin  = 1
+	processorLimitMax  = 10000
+	processorWeightMin = 0
+	processorWeightMax = 10000
+)
+
+var errAlreadyClosed = errors.New("the handle has already been closed")
+
+// SetResourceLimits sets resource limits on the job object (cpu, memory, storage).
+func (job *JobObject) SetResourceLimits(ctx context.Context, limits *jobLimits) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	// Go through and check what limits were specified and apply them to the job.
+	if limits.memoryLimitInMB != 0 {
+		var (
+			basicLimitFlags uint32
+			eliInfo         windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION
+		)
+		basicLimitFlags |= windows.JOB_OBJECT_LIMIT_JOB_MEMORY
+		if (limits.memoryLimitInMB << 20) > math.MaxUint64 {
+			return fmt.Errorf("memory limit value of `%d` is invalid", limits.memoryLimitInMB)
+		}
+		eliInfo.JobMemoryLimit = (limits.memoryLimitInMB << 20)
+		eliInfo.BasicLimitInformation.LimitFlags = basicLimitFlags
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectExtendedLimitInformation, uintptr(unsafe.Pointer(&eliInfo)), uint32(unsafe.Sizeof(eliInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set extended limit info on job object: %s", err)
+		}
+	}
+
+	if limits.cpuRate != 0 || limits.cpuWeight != 0 || limits.cpuCount != 0 {
+		cpuInfo := winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION{}
+		if limits.cpuCount != 0 {
+			numProcessors := uint32(processorinfo.ProcessorCount())
+			if limits.cpuCount > numProcessors {
+				limits.cpuCount = numProcessors
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = calculateJobCPURate(limits.cpuCount, numProcessors)
+		} else if limits.cpuWeight != 0 {
+			if limits.cpuWeight < processorWeightMin || limits.cpuWeight > processorWeightMax {
+				return fmt.Errorf("processor weight value of `%d` is invalid", limits.cpuWeight)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_WEIGHT_BASED
+			cpuInfo.Value = calculateJobCPUWeight(limits.cpuWeight)
+		} else if limits.cpuRate != 0 {
```

Comment:
I think maybe `cpuLimit` would be a better name than `cpuRate`. The current name doesn't really sound like a hard limit on CPU usage to me.

Reply:
Rate is just what it's referred to in all of the job object calls so I went with it instead of what we tend to call it in HCS. I can change if you'd prefer

Reply:
"Rate" is kind of overloaded in the OS call, since it has to refer to several different concepts (weight-based, hard cap, min/max). I'd say let's pick a clearer name here, since we don't have that same constraint.

Reply:
Gotcha, ill change to limit 👍